src/main/scala/strawman/collection/immutable/Vector.scala

```diff
@@ -0,0 +1,1143 @@
+    def hasNext: Boolean = 0 < i
+    def next(): A =
+      if (0 < i) {
+        i -= 1
+        self(i)
+      } else Iterator.empty.next()
+  }
+
+  // Ideally, clients will inline calls to map all the way down, including the iterator/builder methods.
+  // In principle, escape analysis could even remove the iterator/builder allocations and do it
+  // with local variables exclusively. But we're not quite there yet ...
+
+  def apply(index: Int): A = {
+    val idx = checkRangeConvert(index)
+    getElem(idx, idx ^ focus)
+  }
+
+  private def checkRangeConvert(index: Int) = {
+    val idx = index + startIndex
+    if (index >= 0 && idx < endIndex)
+      idx
+    else
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  /*override*/ def updated[B >: A](index: Int, elem: B): Vector[B] = updateAt(index, elem)
+
+  /** Alias for `appendFront`.
+    *
+    * Note that :-ending operators are right associative (see example).
+    * A mnemonic for `+:` vs. `:+` is: the COLon goes on the COLlection side.
+    */
+  @`inline` def +: [B >: A](elem: B): Vector[B] = appendFront(elem)
+
+  /** Alias for `appendBack`
+    *
+    * Note that :-ending operators are right associative (see example).
+    * A mnemonic for `+:` vs. `:+` is: the COLon goes on the COLlection side.
+    */
+  @`inline` def :+ [B >: A](elem: B): Vector[B] = appendBack(elem)
+
+  override def take(n: Int): Vector[A] = {
+    if (n <= 0)
+      Vector.empty
+    else if (startIndex < endIndex - n)
+      dropBack0(startIndex + n)
+    else
+      this
+  }
+
+  override def drop(n: Int): Vector[A] = {
+    if (n <= 0)
+      this
+    else if (startIndex < endIndex - n)
+      dropFront0(startIndex + n)
+    else
+      Vector.empty
+  }
+
+  /*override*/ def takeRight(n: Int): Vector[A] = {
+    if (n <= 0)
+      Vector.empty
+    else if (endIndex - n > startIndex)
+      dropFront0(endIndex - n)
+    else
+      this
+  }
+
+  /*override*/ def dropRight(n: Int): Vector[A] = {
+    if (n <= 0)
+      this
+    else if (endIndex - n > startIndex)
+      dropBack0(endIndex - n)
+    else
+      Vector.empty
+  }
+
+  override def head: A = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.head")
+    apply(0)
+  }
+
+  override def tail: Vector[A] = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.tail")
+    drop(1)
+  }
+
+  override def last: A = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.last")
+    apply(length - 1)
+  }
+
+  /*override*/ def init: Vector[A] = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.init")
+    dropRight(1)
+  }
+
+  override def slice(from: Int, until: Int): Vector[A] =
+    take(until).drop(from)
+
+  override def splitAt(n: Int): (Vector[A], Vector[A]) = (take(n), drop(n))
+
+  // concat (suboptimal but avoids worst performance gotchas)
+  override def concat[B >: A](that: IterableOnce[B]): Vector[B] =
+    that match {
+      case it: Iterable[B] =>
```

Comment:
I have to perform this runtime test in order to apply the `concat` optimizations. An alternative solution would be to change the type signature of `concat` (in `Iterable`) to take an `Iterable` as parameter instead of an `IterableOnce`.